### ISSUE TITLE
Pre-trained segmentation model giving wrong outputs

### ISSUE THREAD

**Issue by tauseef09** [open-mmlab/mmsegmentation#1646]
### **The Issue**

Hi. I was trying to load the pre-trained weights for KNet into the model which I seemed to have done correctly. But the segmentation result are totally wrong. The model just outputs a segmentation map consisting of a single color.

I built the segmentor as such:

```
knet = build_segmentor(model).to(device=DEVICE)
```
I downloaded the model weights (Method: KNet + DeepLabV3) from the table in this [link](https://github.com/open-mmlab/mmsegmentation/tree/master/configs/knet) and then loaded the model with pretrained weights as shown below. I got no errors. Just a message saying **load checkpoint from local path: /content/drive/MyDrive/knet_s3_deeplabv3_r50-d8_8x2_512x512_adamw_80k_ade20k_20220228_041642-00c8fbeb.pth** which seems to me as the model was loaded with the pretrained weights correctly.

```
from mmcv.runner import load_checkpoint
pretrained_path = '/content/drive/MyDrive/knet_s3_deeplabv3_r50-d8_8x2_512x512_adamw_80k_ade20k_20220228_041642-00c8fbeb.pth'
checkpoint = load_checkpoint(knet, pretrained_path, map_location=DEVICE)
knet.CLASSES = checkpoint['meta']['CLASSES']
knet.PALETTE = checkpoint['meta']['PALETTE']
```

The **config for the model** was also downloaded from the same link. I am attaching the model config here for convenience:

```
# model settings
norm_cfg = dict(type='BN', requires_grad=True)
num_stages = 3
conv_kernel_size = 1
model = dict(
    type='EncoderDecoder',
    init_cfg='open-mmlab://resnet50_v1c',
    backbone=dict(
        type='ResNetV1c',
        depth=50,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        dilations=(1, 1, 2, 4),
        strides=(1, 2, 1, 1),
        norm_cfg=norm_cfg,
        norm_eval=False,
        style='pytorch',
        contract_dilation=True),
    decode_head=dict(
        type='IterativeDecodeHead',
        num_stages=num_stages,
        kernel_update_head=[
            dict(
                type='KernelUpdateHead',
                num_classes=150,
                num_ffn_fcs=2,
                num_heads=8,
                num_mask_fcs=1,
                feedforward_channels=2048,
                in_channels=512,
                out_channels=512,
                dropout=0.0,
                conv_kernel_size=conv_kernel_size,
                ffn_act_cfg=dict(type='ReLU', inplace=True),
                with_ffn=True,
                feat_transform_cfg=dict(
                    conv_cfg=dict(type='Conv2d'), act_cfg=None),
                kernel_updator_cfg=dict(
                    type='KernelUpdator',
                    in_channels=256,
                    feat_channels=256,
                    out_channels=256,
                    act_cfg=dict(type='ReLU', inplace=True),
                    norm_cfg=dict(type='LN'))) for _ in range(num_stages)
        ],
        kernel_generate_head=dict(
            type='ASPPHead',
            in_channels=2048,
            in_index=3,
            channels=512,
            dilations=(1, 12, 24, 36),
            dropout_ratio=0.1,
            num_classes=150,
            norm_cfg=norm_cfg,
            align_corners=False,
            loss_decode=dict(
                type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0))),
    auxiliary_head=dict(
        type='FCNHead',
        in_channels=1024,
        in_index=2,
        channels=256,
        num_convs=1,
        concat_input=False,
        dropout_ratio=0.1,
        num_classes=150,
        norm_cfg=norm_cfg,
        align_corners=False,
        loss_decode=dict(
            type='CrossEntropyLoss', use_sigmoid=False, loss_weight=0.4)),
    # model training and testing settings
    train_cfg=dict(),
    test_cfg=Test_cfg(mode='whole', stride=1))
```

The only thing that was modified was the test_cfg creation. I created a class called Test_cfg and initialized it with the attributes 'mode' and 'stride' as I saw the source code trying to access these attributes with a dot (.) operator. The class was declared as below:
```
class Test_cfg():
  def __init__(self, mode='whole', stride=1):
    self.mode = mode
    self.stride = stride 
test_cfg = Test_cfg(mode='whole', stride=1)
```

The dataset in question is ADE20K.

I loaded the input image and also put it through the same test pipeline as the authors did.
```
input_img_filename = '/content/drive/MyDrive/ADE_train_00011900.jpg'
input_img = Image.open(input_img_filename).convert("RGB")
input_img = np.array(input_img)

from albumentations.pytorch import ToTensorV2
import albumentations as A

test_transform = A.Compose(
        [
            A.Resize(height=512, width=512),
            A.Normalize(
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
            ),
            ToTensorV2(),   
        ],
    )

augmentations = test_transform(image=input_img)
input_img_aug = augmentations["image"]
input_img_aug = input_img_aug.to(device=DEVICE)
input_img_aug = input_img_aug.unsqueeze(0)
print(input_img_aug.shape)
```
```
Output: torch.Size([1, 3, 512, 512])
```

And finally obtained the results using the following piece of code:
```
img_meta = [dict(
        img_shape=(512, 512, 3),
        scale_factor=1,
        flip=False,
        ori_shape=(512, 512, 3)
        )]

knet.eval()
preds = knet.inference(input_img_aug, img_meta, rescale=False)
```

The results were displayed using mmseg's built in api:

```
from mmseg.apis import show_result_pyplot

input_img_aug = input_img_aug.to('cpu')
preds = preds.to('cpu')
input_img_aug = input_img_aug.squeeze(0)
input_img_aug = torch.permute(input_img_aug, (1, 2, 0))
preds = preds.squeeze(0)

show_result_pyplot(knet, np.array(input_img_aug), preds, knet.PALETTE, fig_size=(7, 5))
```

### **Results**

These is the image that I want my model to predict the result of, and the corresponding segmentation mask:

![ADE_train_00011900](https://user-images.githubusercontent.com/55648131/172582328-6e0d7568-98e3-4294-b7eb-97f36db5a98e.jpg)
![ADE_train_00011900_seg](https://user-images.githubusercontent.com/55648131/172582336-2fa66462-817a-4921-a9b8-d39aa3e1a200.png)

And here is what my model predicted:

![Capture](https://user-images.githubusercontent.com/55648131/172582543-19e6863d-0c0e-4a75-99fd-005c6c476c92.PNG)

Any kind of assistance will be a huge help. Thanks!

**Comment by MeowZheng** [comment]
> The only thing that was modified was the test_cfg creation. I created a class called Test_cfg and initialized it with the attributes 'mode' and 'stride' as I saw the source code trying to access these attributes with a dot (.) operator. The class was declared as below:

I think there might be a problem that you revised `test_cfg`, Type of `test_cfg` here is [`Config`](https://github.com/open-mmlab/mmcv/blob/834d5978dfbbb9e6797a4524e4de37f3d3de6cd6/mmcv/utils/config.py#L72) from MMCV

If you would like to use `whole_inference`, just try this 
```
test_cfg=dict(mode='whole')
```

**Comment by tauseef09** [comment]
@MeowZheng I tried both ways. Using `test_cfg=dict(mode='whole')` raises me an error saying `'dict' object has no attribute 'mode'` when I try to run `knet.inference(input_img_aug, img_meta, rescale=False)` which makes sense since a python dictionary cannot be accessed using the dot (.) operator which the source code of `knet.inference(input_img_aug, img_meta, rescale=False)` is trying to do.

On the other hand, defining test_cfg as `test_cfg=Config(dict(mode='whole'))`, where [Config](https://github.com/open-mmlab/mmcv/blob/834d5978dfbbb9e6797a4524e4de37f3d3de6cd6/mmcv/utils/config.py#L72) is that of MMCV, gives me the same wrong segmentation output as mentioned in the original problem:

![Capture](https://user-images.githubusercontent.com/55648131/172655134-0055780c-13aa-4218-bbc7-a9fa9faed14d.PNG)

**Comment by tauseef09** [comment]
I think I've solved the issue. For some reason trying to infer an image using the pre-trained KNet model by normalizing the image with the mean and standard deviation along each channel following the author's method doesn't seem to work. So, I removed that part from the test transformation pipeline. Also I needed to add `max_pixel_value=255.0` to the albumentations test pipeline. So, the final test transformation was:
```
test_transform = A.Compose(
        [
            A.Resize(height=512, width=512),
            A.Normalize(
                max_pixel_value=255.0,
            ),
            ToTensorV2(),   
        ],
    )
```

Another point to note is that mmseg.api's `show_result_pyplot()` function doesn't seem to output the correct label. So, I created my own function that converts predictions to rgb images. This function only converts the segmentation map predicted by the model to an rgb image which can be displayed by pyplot. 
```
def onehot_to_img(predicted_one_hot, color_palette):   
    one_hot_np = np.array(predicted_one_hot)
    img = np.zeros((one_hot_np.shape[0], one_hot_np.shape[1], 3), dtype=np.uint8)
    for h in range(one_hot_np.shape[0]):
        for w in range(one_hot_np.shape[1]):
            idx = one_hot_np[h][w]
            img[h][w] = list(color_palette[idx])
    return img
```

Then I used this function as shown below to get my final output:
```
pred_seg_mask = torch.argmax(preds, dim=1)
pred_seg_mask = pred_seg_mask.to("cpu")
pred_seg_mask = onehot_to_img(pred_seg_mask[0], knet.PALETTE)
plt.imshow(pred_seg_mask)
```
I got my desired output segmentation map which looked like this:

![Capture2](https://user-images.githubusercontent.com/55648131/172832079-5207f71e-262e-4e5e-ab7a-f0a018fcf5f5.PNG)
